### ISSUE TITLE
User Story issue

### ISSUE THREAD

**Issue by cmessick** [scimma/hop-client#189]
### Description
As a user, I want to be able to specify the delivery_callback function when I set up the producer. 

Currently delivery callbacks can only be specified at the time of writing. It's the only configuration related item I set at the time of writing a message, it would be nice organizationally to set it at the time I instantiate the writable stream object instead of every time I write to it. 

I poked around in the documentation of `confluent-kafka`, and it looks like this can be set with the `dr_cb` config variable, just in case that helps at all.

**Comment by cnweaver** [comment]
This would be possible, but I think I would like to understand the use case a bit better to try to design an implementation. 

For reference, we use the delivery callback to make the `RobustProducer` work, and for that it's important to set it per-message, as otherwise we wouldn't know for which message our callback was being invoked: https://github.com/scimma/hop-client/blob/master/hop/robust_publisher.py#L691 . Is there a use-case for a delivery callback without distinguishing the particular message?

**Comment by cmessick** [comment]
My understanding is that it's purely aesthetics, though @lpsinger should chime in if he disagrees.

My understanding of delivery callbacks though is that the delivery callback function returns both the message and the error (if any), so I don't think I understand why you don't know which message is invoking the callback if specifying it when setting up the producer as opposed to setting it when you send a message.

**Comment by cnweaver** [comment]
Getting back the full information about the message in question is not implemented in the python interface: https://github.com/confluentinc/confluent-kafka-python/pull/1167 . I ran into this because I wanted to use headers to attach a unique identifier to messages, since otherwise the end-user might send two or more identical messages, which my code wouldn't be able to distinguish. However, if you know your messages will have distinct/recognizable bodies, I guess it could work for you as-is.

**Comment by cmessick** [comment]
Yeah I believe my messages will have distinct / recognizable enough bodies, at the very least we include the message generation time so we have guaranteed uniqueness (we'll never send messages fast enough to need high precision timestamps).

**Comment by lpsinger** [comment]
Correct, we don't care about the headers for our application.
